### ISSUE TITLE
The scraper omits some advisors

### ISSUE THREAD

**Issue by david-wahlstedt** [j2kun/math-genealogy-scraper#5]
I cloned this repo, installed and run it: nice program!
The latest commit in my clone is 89e29cb9.

Issue: Some ancestors are missing when the data is collected/parsed.

Two examples:

### 1
[Gérard Huet](https://genealogy.math.ndsu.nodak.edu/id.php?id=83536) had two supervisors(and did two doctorate graduations) , but the scraper only includes the first one.
info.csv, line 97497:
```
83536,Gérard Pierre Huet,1972
```
graph.csv, line 97497: 
```
83536,136510,137176 137173 137169 137164 137175 83529 137163 137166 137167 137161 95088 137165 137162 137170 137160 137172 137174 128359 140162 137168 137177 137171
``` 
After 136510 the second advisor (76348) is missing, but he is present in the info.csv:
```
76348,Maurice Paul Nivat,1967
```
### 2
[Shams al‐Dīn al‐Bukhārī](https://genealogy.math.ndsu.nodak.edu/id.php?id=204293) did have one advisor according to the database, but the scraper doesn't include him.
info.csv, line 131729:
```
204293,Shams ad-Din Al-Bukhari,None
```
graph.csv, line 131729:
```
204293,,201288
```
No advisor included, and this one is not in info.csv.
According to the database he was [Nasir al-Dīn al-Ṭūsī](https://genealogy.math.ndsu.nodak.edu/id.php?id=217509)(number 217509)
Maybe the parser is sensitive to the unusual characters?

### 

I looked at parse.py, line 73, maybe it's somewhere near that?

Best regards,
David

**Comment by j2kun** [comment]
I see the issue with the multiple advisors. I assume that multiple advisors are in the same container when they are actually in different containers:

https://github.com/j2kun/math-genealogy-scraper/blob/master/parse.py#L72-L76

I think Nasir al-Dīn al-Ṭūsī was added within the last year, because I don't have any records of ids larger than 210000. I will fix this multiple advisor issue, re-run, and reupload a new scrape, checking for Nasir's entry.

**Comment by david-wahlstedt** [comment]
Great, thank you!

**Comment by j2kun** [comment]
Confirmed both the issues are fixed in the most recent scrape. The `genealogy_graph.json` file contains the most up-to-date version of the graph, and has two edges for the multiple advisors. The following `jq` query shows it:

```
$  jq '.edges[] | select(.[1] == 83536)' genealogy_graph.json 
[
  76348,
  83536
]
[
  136510,
  83536
]
```

Nasir is in there too:

```
$ grep 217509 info.csv 
217509, Nasir al-Dīn al-Ṭūsī, null
```

**Comment by david-wahlstedt** [comment]
Thanks!
I downloaded the new version, and it looks now like it's repeating some advisor links:
to be able to grep: (line breaks would make it easier to grep manually)
```
tr \} "\n">d.l < data.json 

grep id\"\:\ 242176, d.l
, {"id": 242176, "name": "Urban Larsson", "thesis": "Impartial games and recursive functions", "school": "Chalmers Tekniska H\u00f6gskola", "country": "Sweden", "year": 2013, "subject": "05\u2014Combinatorics", "advisors": [75373, 53169, 75373, 53169], "students": []
```
with jq this looks as follows:
```
jq '.edges[] | select(.[1] == 242176)' genealogy_graph.json 
[
  53169,
  242176
]
[
  75373,
  242176
]
```
so the repetitions aren't visible there.
These repetitions are not present in data.json for id 83536.

Best regards,
David

**Comment by david-wahlstedt** [comment]
I found records where all fields except the id are null, for instance id=193664. These are searchable from the math genealogy database, and yield "Unknown",etc. These are present in the data.json file.

**Comment by j2kun** [comment]
Hmmm at some point it would be nice to add proper testing...

Anyhow, I added commit f8c9a96b198933801aea86c3afda1270360975bd which should fix the duplicate advisor issue. I feel bad running another re-scrape of the entire database to fix it, so instead I used jq

```
jq '{"nodes": [.nodes[] | .advisors? |= unique]}' data.json > data_uniq.json
```

Also note that `jq '.' data.json` will format a json file nicely for use with grep.

Not sure what to do about the "Unknowns," seems like bad data, but shouldn't be a major problem in itself.

I also migrated the big files to git-lfs.
